pkg/querier/replication.go

```diff
@@ -46,26 +73,300 @@ func forGivenReplicationSet[Result any, Querier any](ctx context.Context, client
+					shardsSeen = make([]bool, shards)
+				} else {
+					shardsSeen = shardsSeen[:shards]
+					for idx := range shardsSeen {
+						shardsSeen[idx] = false
+					}
+				}
+			}
+
+			if len(shardsSeen) != int(shards) {
+				panic("shard length mismatch")
+			}
+
+			shardsSeen[shardIdx] = true
+		}
+		// check if all shards are present
+		allShardsPresent := true
+		for _, shardSeen := range shardsSeen {
+			if !shardSeen {
+				allShardsPresent = false
+				break
+			}
+		}
+
+		if allShardsPresent {
+			continue
+		}
+
+		// now remove all blocks that are shareded but not complete
+		for _, block := range shardedBlocks {
+			r.removeBlock(block)
+		}
+	}
+}
+
+// prunes blocks that are contained by a higher compaction level block
+func (r *replicasPerBlockID) pruneSupersededBlocks() {
+	for blockID := range r.m {
+		meta, ok := r.meta[blockID]
+		if !ok {
+			panic("meta missing")
+		}
+		if meta.Compaction == nil {
+			continue
+		}
+		if meta.Compaction.Level < 2 {
+			continue
+		}
+		for _, blockID := range meta.Compaction.Parents {
+			r.removeBlock(blockID)
+		}
+		for _, blockID := range meta.Compaction.Sources {
+			r.removeBlock(blockID)
+		}
+	}
+}
+
+func (r *replicasPerBlockID) blockPlan() map[string]*ingestv1.BlockHints {
+	var (
+		deduplicate = false
+		hash        = xxhash.New()
+		plan        = make(map[string]*ingestv1.BlockHints)
+	)
+
+	r.pruneIncompleteShardedBlocks()
+	r.pruneSupersededBlocks()
+
+	// now we go through all blocks and choose the replicas that we want to query
+	for blockID, replicas := range r.m {
+		// skip if we have no replicas, then block is already contained i an higher compaction level one
+		if len(replicas) == 0 {
+			continue
+		}
+
+		meta, ok := r.meta[blockID]
+		if !ok {
+			continue
+		}
+		// when we see a single ingester block, we want to deduplicate
+		if meta.Compaction != nil && meta.Compaction.Level == 1 {
+			deduplicate = true
+		}
+
+		// only get store gateways replicas
+		sgReplicas := lo.Filter(replicas, func(replica string, _ int) bool {
+			t, ok := r.instanceType[replica]
+			if !ok {
+				return false
+			}
+			return t == storeGatewayInstance
+		})
+
+		if len(sgReplicas) > 0 {
+			// if we have store gateway replicas, we want to query them
+			replicas = sgReplicas
+		}
+
+		// now select one replica, based on block id
+		sort.Strings(replicas)
+		hash.Reset()
+		_, _ = hash.WriteString(blockID)
+		hashIdx := int(hash.Sum64())
+		if hashIdx < 0 {
+			hashIdx = -hashIdx
+		}
+		selectedReplica := replicas[hashIdx%len(replicas)]
+
+		// add block to plan
+		p, exists := plan[selectedReplica]
+		if !exists {
+			p = &ingestv1.BlockHints{}
+			plan[selectedReplica] = p
+		}
+		p.Ulids = append(p.Ulids, blockID)
+
+		// set the selected replica
+		r.m[blockID] = []string{selectedReplica}
+	}
+
+	// adapt the plan to make sure all replicas will deduplicate
+	if deduplicate {
+		for _, hints := range plan {
+			hints.Deduplication = deduplicate
+		}
+	}
```

Comment:
Does it mean that if we got a query like "last 2 days" we will force deduplication as the plan also includes blocks that have not been compacted yet? I guess, this should be mitigated by query time range split – I'm wondering if we should set to the block duration by default

Reply:
Yes this is true. The first step is for the query time range split to alight with roughly the processing time for the compacted block. So it should be slightly higher than block duration, just because we introduce a fair bit of latency before a block is compacted/deduplicated and available to store-gateways.

Ideally in a future where the query-frontend will also be aware of the plan, we can split accordingly to what the block layout is.